sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,760 @@
+
+class memory_pool {
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(context ctx, Properties props = {});
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(queue q, usm::alloc kind, Properties props = {});
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(context ctx, void *ptr, size_t size, Properties props = {});
+
+  ~memory_pool();
+
+  context get_context() const;
+
+  device get_device() const;
+
+  usm::alloc get_alloc_kind() const;
+
+  size_t get_max_size() const;
+
+  size_t get_threshold() const;
+
+  void set_new_threshold(size_t newThreshold);
+
+}; // memory_pool
+
+}  // ext::oneapi::experimental
+----
+
+[source, c++]
+----
+template <typename Properties = empty_properties_t>
+memory_pool(context ctx, Properties props = {});
+----
+
+_Effects_: Constructs a memory pool associated with `ctx` and all SYCL devices
+associated with it, with the allocation kind `usm::alloc::host` and applying any
+properties in `props`.
+
+[source, c++]
+----
+template <typename Properties = empty_properties_t>
+memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});
+----
+
+_Effects_: Constructs a memory pool associated with `ctx` and `dev`, with the
+allocation kind `kind` and applying any properties in `props`.
+
+_Throws_: An exception with the `errc::invalid` error code if `kind` is
+`usm::alloc::host`.
+
+[source, c++]
+----
+template <typename Properties = empty_properties_t>
+memory_pool(queue q, usm::alloc kind, Properties props = {});
+----
+
+_Effects_: Constructs a memory pool associated with the SYCL context and device
+from `q` and `dev`, with the allocation kind `kind` and applying any properties
+in `props`. If `kind` is `usm::alloc::host` this is equivelant to calling
+`memory_pool(queue.get_context(), props)` otherwise this is equivelant to
+calling `memory_pool(queue.get_context(), queue.get_device(), props)`.
+
+[source, c++]
+----
+template <typename Properties = empty_properties_t>
+memory_pool(context ctx, void *ptr, size_t size, Properties props = {});
+----
+
+_Effects_: Constructs a memory pool associated with `ctx` which uses an existing
+USM memory allocation `ptr` of size (in bytes) `size`, and applying any
+properties in `props`. The memory pool will use the existing USM memory
+allocation instead of allocating from the SYCL runtime. The associated SYCL
+device and allocation kind of the memory pool is inferred from the `ptr`. If
+`ptr` is not a valid USM pointer of kind `usm::alloc::host`,
+`usm::alloc::device` or `usm::alloc::shared` the result is undefined behavior.
+The maximum size and initial threshold of the memory pool are set to `size`.
+
+_Throws_: An exception with the `errc::invalid` error code if `props` contains
+the `maximum_size` or `initial_threshold` properties.
```

Comment:
I'm not sure I understand why `max_size` and `initial_threshold` are properties at all.

It seems like there are dedicated queries for them (`get_max_size()` and `get_threshold()`).  Without any properties, the max size is set to an implementation-defined limit, and the threshold is set to zero.

Why not just use a regular overload here with some runtime values?
```c++
template <typename Properties = empty_properties_t>
memory_pool(queue q, usm::alloc kind, int initial_threshold = 0, int maximum_size = /* implementation-defined */, Properties props = {});
```

That seems like it would give the best error messages, and it wouldn't pull in the complexities of using properties.